lab_11.cpp

```diff
@@ -0,0 +1,200 @@
+
+      // Segment out the areas of the image that fits well enough.
+      cv::Mat segmented = performSegmentation(mahalanobis_img, current_threshold, use_otsu);
+      cv::setTrackbarPos(threshold_trackbar_name, win_name_input, current_threshold);
+
+      // Set segmented area to green.
+      frame.setTo(cv::Scalar{0, 255, 0}, segmented);
+    }
+
+    // Update if using adaptive model.
+    if (model && use_adaptive_model)
+    {
+      cv::Mat new_samples = extractTrainingSamples(feature_image, sampling_rectangle);
+      updateSamples(current_samples, new_samples, adaptive_update_ratio);
+      model = std::make_shared<MultivariateNormalModel>(current_samples);
+    }
+
+    // Draw current frame.
+    drawSamplingRectangle(frame, sampling_rectangle);
+    cv::imshow(win_name_input, frame);
+
+    // Get input from keyboard
+    char key = static_cast<char>(cv::waitKey(10));
+    if (key == ' ')
+    {
+      // Press space to train a model based on the samples in the rectangle.
+      current_samples = extractTrainingSamples(feature_image, sampling_rectangle);
+      model = std::make_shared<MultivariateNormalModel>(current_samples);
+    }
+    else if (key == 'o')
+    {
+      // Press 'o' to toggle use of Otsu's method.
+      use_otsu = !use_otsu;
+    }
+    else if (key == 'a')
+    {
+      // Press 'a' to toggle use of adaptive model.
+      use_adaptive_model = !use_adaptive_model;
+    }
+    else if (key >= 0)
+    {
+      // Press any other key to exit the lab.
+      break;
+    }
+  }
+}
+
+
+void updateSamples(cv::Mat& old_samples, const cv::Mat& new_samples, float update_ratio)
+{
+  // TODO 4: Implement a random update of samples given the ratio of new_samples
+  old_samples = new_samples; // Dummy, replace
+}
+
+
+cv::Mat extractFeatures(const cv::Mat& frame)
+{
+  // TODO 5: Extract other/more features for each pixel.
+  cv::Mat feature_image = frame.clone();
+  return feature_image;
+}
+
+
+cv::Mat performSegmentation(const cv::Mat& input_image, int& threshold_value, bool use_otsu)
+{
+  // TODO 6: Add morphological operations to reduce noise, and other fancy segmentation approaches.
+  int thresh_type = cv::THRESH_BINARY_INV;
+  if (use_otsu)
+  { thresh_type |= cv::THRESH_OTSU; }
+
+  cv::Mat segmented_image;
+  threshold_value = static_cast<int>(cv::threshold(input_image, segmented_image, threshold_value, 255, thresh_type));
+
+  return segmented_image;
+}
+
+
+cv::Rect getSamplingRectangle(const cv::Size& img_size)
+{
+  int center_x = img_size.width/2;
+  int center_y = 4*img_size.height/5;
+  int width = 100;
+  int height = 80;
+  int top_left_x = center_x - width/2;
+  int top_left_y = center_y - height/2;
+
+  const cv::Rect sampling_rectangle(top_left_x, top_left_y, width, height);
+  const cv::Rect entire_image(0,0,img_size.width,img_size.height);
+
+  // This operation ensures that the boundardies of the returned sampling rectange is within the image's dimentions,
```

Comment:
boundaries

Reply:
Eh, det heter boundardies... På fransk! 😇 

PS, du kan legge inn suggestion med den her: 
<img width="67" alt="image" src="https://user-images.githubusercontent.com/11663739/106820383-8860fc00-667b-11eb-8b0b-1e6b956fb3d2.png">, så er det bare å klikke for å godkjenne.

Reply:
```suggestion
  // This operation ensures that the boundaries of the returned sampling rectange is within the image's dimentions,
```

Reply:
Nå la jeg også merke til "rectange"...

Reply:
Hvor finner jeg knappen for å legge inn suggestion egentlig? Jeg ser den ikke i tilknytning til denne tekstboksen som dukker opp når jeg trykker på "+" i dokumentet.

Jeg valgte "Add single comment" for denne tråden. Burde jeg valgt "Start a review" istendenfor?

Reply:
Dere gikk begge glipp av "dimentions"...

Reply:
Altså, "dimentions" står i ordboka, bare se...

<img width="241" alt="image" src="https://user-images.githubusercontent.com/11663739/106861082-90458e00-66c5-11eb-8200-91e6c0960d28.png">